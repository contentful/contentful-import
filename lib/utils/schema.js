import Joi from 'joi'

const entrySchema = {
  sys: Joi.object(),
  fields: Joi.object()
}

const contentTypeSchema = {
  sys: Joi.object(),
  fields: Joi.array().required().items(Joi.object().keys({
    id: Joi.string().required(),
    name: Joi.string().required(),
    type: Joi.string().required().regex(/^Symbol|Text|Integer|Number|Date|Object|Boolean|Array|Link|Location$/),
    validations: Joi.array(),
    disabled: Joi.boolean(),
    omitted: Joi.boolean(),
    required: Joi.boolean(),
    localized: Joi.boolean(),
    linkType: Joi.string().when('type', {is: 'Link', then: Joi.string().regex(/^Asset|Entry$/), otherwise: Joi.forbidden()})
  }))
}

const assetSchema = {
  sys: Joi.object(),
  fields: Joi.object({
    file: Joi.object().pattern(/.+/, Joi.object({
      url: Joi.string().required(),
      details: Joi.object({
        size: Joi.number(),
        image: Joi.object({
          width: Joi.number(),
          height: Joi.number()
        })}),
      fileName: Joi.string().required(),
      contentType: Joi.string().required()
    }))
  }).required()
}
const editorInterfaceSchema = {
  sys: Joi.object(),
  controls: Joi.array().items([{fieldId: Joi.string(), widgetId: Joi.string()}])
}
const localeSchema = {
  name: Joi.string().required(),
  internal_code: Joi.string(),
  code: Joi.string().required(),
  fallbackCode: Joi.string().allow([null]),
  default: Joi.boolean(),
  contentManagementApi: Joi.boolean(),
  contentDeliveryApi: Joi.boolean(),
  optional: Joi.boolean(),
  sys: Joi.object()
}

const webhookSchema = {
  name: Joi.string(),
  url: Joi.string().uri().required(),
  topics: Joi.array().required(),
  httpBasicUsername: Joi.string()
}

const payloadSchema = Joi.object({
  entries: Joi.array().items([entrySchema]),
  contentTypes: Joi.array().items([contentTypeSchema]),
  assets: Joi.array().items([assetSchema]),
  locales: Joi.array().items([localeSchema]),
  editorInterfaces: Joi.array().items([editorInterfaceSchema]),
  webhooks: Joi.array().items([webhookSchema])
})
export {
  entrySchema,
  contentTypeSchema,
  localeSchema,
  webhookSchema,
  editorInterfaceSchema,
  assetSchema,
  payloadSchema
}
