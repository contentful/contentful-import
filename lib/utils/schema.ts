import Joi from 'joi'

const entrySchema = {
  sys: Joi.object(),
  fields: Joi.object()
}

const tagSchema = {
  name: Joi.string().required(),
  sys: Joi.object()
}

const contentTypeSchema = {
  sys: Joi.object(),
  fields: Joi.array().required().items(Joi.object().keys({
    id: Joi.string().required(),
    name: Joi.string().required(),
    type: Joi.string().required().regex(/^Symbol|Text|Integer|Number|Date|Object|Boolean|Array|Link|Location$/),
    validations: Joi.array(),
    disabled: Joi.boolean(),
    omitted: Joi.boolean(),
    required: Joi.boolean(),
    localized: Joi.boolean(),
    linkType: Joi.string().when('type', { is: 'Link', then: Joi.string().regex(/^Asset|Entry$/), otherwise: Joi.forbidden() })
  }))
}

const assetSchema = {
  sys: Joi.object(),
  fields: Joi.object({
    file: Joi.object().pattern(/.+/, Joi.object({
      url: Joi.string().required(),
      details: Joi.object({
        size: Joi.number(),
        image: Joi.object({
          width: Joi.number(),
          height: Joi.number()
        })
      }),
      fileName: Joi.string().required(),
      contentType: Joi.string().required()
    }))
  }).required()
}
const editorInterfaceSchema = {
  sys: Joi.object(),
  controls: Joi.array().items({ fieldId: Joi.string(), widgetId: Joi.string() })
}
const localeSchema = {
  name: Joi.string().required(),
  internal_code: Joi.string(),
  code: Joi.string().required(),
  fallbackCode: Joi.string().allow(null),
  default: Joi.boolean(),
  contentManagementApi: Joi.boolean(),
  contentDeliveryApi: Joi.boolean(),
  optional: Joi.boolean(),
  sys: Joi.object()
}

const webhookSchema = {
  name: Joi.string(),
  url: Joi.string().replace(/{[^}{]+?}/g, 'x').regex(/^https?:\/\/[^ /}{][^ }{]*$/i).required(),
  topics: Joi.array().required(),
  httpBasicUsername: Joi.string().allow('', null)
}

/**
 * @returns normalized validation object. Don't use normalized output as payload
 */
const payloadSchema = Joi.object({
  entries: Joi.array().items(entrySchema),
  contentTypes: Joi.array().items(contentTypeSchema),
  tags: Joi.array().items(tagSchema),
  assets: Joi.array().items(assetSchema),
  locales: Joi.array().items(localeSchema),
  editorInterfaces: Joi.array().items(editorInterfaceSchema),
  webhooks: Joi.array().items(webhookSchema)
})
export {
  entrySchema,
  contentTypeSchema,
  tagSchema,
  localeSchema,
  webhookSchema,
  editorInterfaceSchema,
  assetSchema,
  payloadSchema
}
