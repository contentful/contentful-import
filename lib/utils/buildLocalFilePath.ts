import path from 'path'

export function buildLocalFilePath (url, directory, fileName) {
  // handle urls without protocol
  if (url.startsWith('//')) {
    url = 'https:' + url
  }

  const { host, pathname } = new URL(url)
  // Extract directory path without leading slash
  const pathWithoutFilename = pathname.substring(1, pathname.lastIndexOf('/') + 1)
  const localFilePath = fileName ? path.join(directory, host, pathWithoutFilename, fileName) : path.join(directory, host, decodeURIComponent(pathname))

  return localFilePath
}
