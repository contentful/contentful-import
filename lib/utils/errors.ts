import type { AssetProps, EditorInterfaceProps, EntryProps } from 'contentful-management'

export class ContentfulAssetError extends Error {
  filePath: string
  constructor (message: string, filePath: string) {
    super(message)
    this.filePath = filePath
  }
}

export class ContentfulValidationError extends Error {
  error?: {
    sys: { id: string };
    details: {
      errors: Array<{ name: string }>;
    };
  }
}

export class ContentfulEntityError extends Error {
  entity: EntryProps | AssetProps | EditorInterfaceProps
}

export class ContentfulMultiError extends Error {
  errors: ContentfulValidationError[]
}
