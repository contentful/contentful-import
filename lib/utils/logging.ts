import { LogMessage } from 'contentful-batch-libs'

type ErrorMessage = Extract<LogMessage, { level: 'error' }>

export function isErrorLog (log: LogMessage): log is ErrorMessage {
  return log.level !== 'info' && log.level !== 'warning'
}

export function isDisplayLog (log: LogMessage): boolean {
  return log.level !== 'info'
}
