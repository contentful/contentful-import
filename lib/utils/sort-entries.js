import { some, filter, map } from 'lodash/collection'
import * as _o from 'lodash/object'
import { flatten } from 'lodash/array'

/**
 * Given a list of entries, this function reorders them so that entries which
 * are linked from other entries always come first in the order. This ensures
 * that when we publish entries, we are not publishing entries which contain
 * links to other entries which haven't been published yet.
 */
export default function sortEntries (entries) {
  const linkedEntries = getLinkedEntries(entries)

  const mergedLinkedEntries = mergeSort(linkedEntries, (a) => {
    return hasLinkedIndexesInFront(a)
  })

  return map(mergedLinkedEntries, (linkInfo) => entries[linkInfo.index])

  function hasLinkedIndexesInFront (item) {
    if (hasLinkedIndexes(item)) {
      return some(item.linkIndexes, (index) => index > item.index) ? 1 : -1
    }
    return 0
  }

  function hasLinkedIndexes (item) {
    return item.linkIndexes.length > 0
  }
}

function getLinkedEntries (entries) {
  return map(entries, function (entry) {
    const entryIndex = entries.indexOf(entry)

    const rawLinks = map(entry.fields, (field) => {
      field = _o.values(field)[0]
      if (isEntryLink(field)) {
        return getFieldEntriesIndex(field, entries)
      } else if (isEntityArray(field) && isEntryLink(field[0])) {
        return map(field, (item) => getFieldEntriesIndex(item, entries))
      }
    })

    return {
      index: entryIndex,
      linkIndexes: filter(flatten(rawLinks), (index) => index >= 0)
    }
  })
}

function getFieldEntriesIndex (field, entries) {
  const id = _o.get(field, 'sys.id')
  return entries.findIndex((entry) => entry.sys.id === id)
}

function isEntryLink (item) {
  return _o.get(item, 'sys.type') === 'Entry' ||
  _o.get(item, 'sys.linkType') === 'Entry'
}

function isEntityArray (item) {
  return Array.isArray(item) && item.length > 0 && _o.has(item[0], 'sys')
}

/**
 * From https://github.com/millermedeiros/amd-utils/blob/master/src/array/sort.js
 * MIT Licensed
 * Merge sort (http://en.wikipedia.org/wiki/Merge_sort)
 * @version 0.1.0 (2012/05/23)
 */
function mergeSort (arr, compareFn) {
  if (arr.length < 2) return arr

  if (compareFn == null) compareFn = defaultCompare

  const mid = ~~(arr.length / 2)
  const left = mergeSort(arr.slice(0, mid), compareFn)
  const right = mergeSort(arr.slice(mid, arr.length), compareFn)

  return merge(left, right, compareFn)
}

function defaultCompare (a, b) {
  return a < b ? -1 : (a > b ? 1 : 0)
}

function merge (left, right, compareFn) {
  var result = []

  while (left.length && right.length) {
    if (compareFn(left[0], right[0]) <= 0) {
      // if 0 it should preserve same order (stable)
      result.push(left.shift())
    } else {
      result.push(right.shift())
    }
  }

  if (left.length) result.push.apply(result, left)
  if (right.length) result.push.apply(result, right)

  return result
}
