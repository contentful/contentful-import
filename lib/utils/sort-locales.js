/**
 * Given a list of locales, this utility will sort them by `fallback` order, i.e. the locales without fallbacks first,
 * then the locales having them as fallbacks, and then recursively.
 */
export default function sortLocales (locales) {
  const localeByFallback = {}

  locales.forEach(locale => {
    if (locale.fallbackCode === null) {
      locale.fallbackCode = undefined
    }

    if (!localeByFallback[locale.fallbackCode]) {
      localeByFallback[locale.fallbackCode] = []
    }

    localeByFallback[locale.fallbackCode].push(locale)
  })

  return sortByFallbackKey(localeByFallback)
}

function sortByFallbackKey (localeByFallback, key) {
  if (!localeByFallback[key]) {
    return []
  }

  const sortedLocales = localeByFallback[key]

  sortedLocales.forEach(locale => {
    sortByFallbackKey(localeByFallback, locale.code)
      .forEach(x => sortedLocales.push(x))
  })

  return sortedLocales
}
