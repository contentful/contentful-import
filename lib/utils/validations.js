import Joi from 'joi'
import { payloadSchema } from './schema'
import getEntityName from 'contentful-batch-libs/dist/get-entity-name'

const attachEntityName = (details, payload) => {
  details.map((detail) => {
    if (detail.path.length >= 2) {
      detail.entity = getEntityName(payload[detail.path[0]][detail.path[1]])
    }
    return detail
  })
}
const countInvalidEntities = (validationData) => {
  const entityCount = validationData.reduce((entities, currentDetail) => {
    if (!entities[currentDetail.path[0]]) {
      entities[currentDetail.path[0]] = 1
    } else {
      entities[currentDetail.path[0]]++
    }
    return entities
  }, {})
  return Object.keys(entityCount).map((key) => `${key}:${entityCount[key]}`)
}

const assertPayload = (payload) => {
  const result = Joi.validate(payload, payloadSchema, {allowUnknown: true, abortEarly: false})
  if (result.error) {
    attachEntityName(result.error.details, payload)
    const invalidEntityCount = countInvalidEntities(result.error.details).join(', ')
    result.error.message = `${invalidEntityCount} - Get further details in the error log file`
    delete result.error._object
    throw result.error
  }
}

const assertDefaultLocale = (source, destination) => {
  const sourceDefaultLocale = source.locales.find((locale) => locale.default === true)
  const destinationDefaultLocale = destination.locales.find((locale) => locale.default === true)

  if (!sourceDefaultLocale || !destinationDefaultLocale) {
    return
  }

  if (sourceDefaultLocale.code !== destinationDefaultLocale.code) {
    throw new Error(`
      Please make sure the destination space have the same default locale as the source\n
      Default locale for source space : ${sourceDefaultLocale.code}\n
      Default locale for destination space: ${destinationDefaultLocale.code}\n
    `)
  }
}

export {
  assertPayload,
  assertDefaultLocale
}
