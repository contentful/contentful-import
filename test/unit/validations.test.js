import { assertPayload } from '../../lib/utils/validations'

test('payload validation should succeed when given an empty payload', () => {
  expect(
    () => assertPayload({entries: [], locales: [], contentTypes: [], assets: []})
  ).not.toThrow()
})

test('payload validation should succeed when given a valid payload', () => {
  const payload = {
    entries: [],
    locales: [],
    contentTypes: [{
      'sys': {
        'space': {
          'sys': {
            'type': 'Link',
            'linkType': 'Space',
            'id': '28p9vvm1oxuw'
          }
        },
        'id': 'person',
        'type': 'ContentType',
        'createdAt': '2017-05-11T12:04:23.540Z',
        'updatedAt': '2017-05-18T14:04:51.650Z',
        'createdBy': {
          'sys': {
            'type': 'Link',
            'linkType': 'User',
            'id': '066RqBikAjzKy0SWUEtFvH'
          }
        },
        'updatedBy': {
          'sys': {
            'type': 'Link',
            'linkType': 'User',
            'id': '2AAFsI4st4sZPlF1LFT13q'
          }
        },
        'publishedCounter': 7,
        'version': 14,
        'publishedBy': {
          'sys': {
            'type': 'Link',
            'linkType': 'User',
            'id': '2AAFsI4st4sZPlF1LFT13q'
          }
        },
        'publishedVersion': 13,
        'firstPublishedAt': '2017-05-11T12:04:23.869Z',
        'publishedAt': '2017-05-18T14:04:51.628Z'
      },
      'displayField': 'name',
      'name': 'Person',
      'description': '',
      'fields': [
        {
          'id': 'location',
          'name': 'Location',
          'type': 'Location',
          'localized': false,
          'required': true,
          'validations': [],
          'disabled': false,
          'omitted': false
        },
        {
          'id': 'name',
          'name': 'Name',
          'type': 'Symbol',
          'localized': false,
          'required': true,
          'validations': [],
          'disabled': false,
          'omitted': false
        },
        {
          'id': 'title',
          'name': 'Title',
          'type': 'Symbol',
          'localized': false,
          'required': true,
          'validations': [],
          'disabled': false,
          'omitted': false
        },
        {
          'id': 'company',
          'name': 'Company',
          'type': 'Symbol',
          'localized': false,
          'required': true,
          'validations': [],
          'disabled': false,
          'omitted': false
        },
        {
          'id': 'shortBio',
          'name': 'Short Bio',
          'type': 'Text',
          'localized': false,
          'required': true,
          'validations': [],
          'disabled': false,
          'omitted': false
        },
        {
          'id': 'email',
          'name': 'Email',
          'type': 'Symbol',
          'localized': false,
          'required': false,
          'validations': [],
          'disabled': false,
          'omitted': false
        },
        {
          'id': 'phone',
          'name': 'Phone',
          'type': 'Symbol',
          'localized': false,
          'required': false,
          'validations': [],
          'disabled': false,
          'omitted': false
        },
        {
          'id': 'facebook',
          'name': 'Facebook',
          'type': 'Symbol',
          'localized': false,
          'required': false,
          'validations': [],
          'disabled': false,
          'omitted': false
        },
        {
          'id': 'twitter',
          'name': 'Twitter',
          'type': 'Symbol',
          'localized': false,
          'required': false,
          'validations': [],
          'disabled': false,
          'omitted': false
        },
        {
          'id': 'github',
          'name': 'Github',
          'type': 'Symbol',
          'localized': false,
          'required': false,
          'validations': [],
          'disabled': false,
          'omitted': false
        },
        {
          'id': 'image',
          'name': 'Image',
          'type': 'Link',
          'localized': false,
          'required': false,
          'validations': [],
          'disabled': false,
          'omitted': false,
          'linkType': 'Asset'
        }
      ]
    }],
    assets: []
  }
  expect(() => assertPayload(payload)).not.toThrow()
})
test('allow null and empty string for webhook httpBasicUsername', () => {
  const webhook = {
    'sys': {
      'type': 'WebhookDefinition',
      'id': '0KzM2HxYr5O1pZ4SaUzK8h',
      'version': 1,
      'space': {
        'sys': {
          'type': 'Link',
          'linkType': 'Space',
          'id': 'yadj1kx9rmg0'
        }
      },
      'createdAt': '2015-05-18T11:29:46.809Z',
      'createdBy': {
        'sys': {
          'type': 'Link',
          'linkType': 'User',
          'id': '7BslKh9TdKGOK41VmLDjFZ'
        }
      },
      'updatedAt': '2015-05-18T11:29:46.809Z',
      'updatedBy': {
        'sys': {
          'type': 'Link',
          'linkType': 'User',
          'id': '4FLrUHftHW3v2BLi9fzfjU'
        }
      }
    },
    'name': 'My webhook',
    'url': 'https://www.example.com',
    'topics': [
      'Entry.create',
      'ContentType.create',
      '*.publish',
      'Asset.*'
    ],
    'httpBasicUsername': 'yolo',
    'headers': [
      {
        'key': 'header1',
        'value': 'value1'
      },
      {
        'key': 'header2',
        'value': 'value2'
      }
    ]
  }
  expect(() => assertPayload({webhooks: [webhook]})).not.toThrow()
})
test('payload validation should fail when given an invalid data', () => {
  const brokenAsset = {
    sys: {id: 'myAsset'},
    fields: {
      title: {'en-US': 'a title'},
      file: {
        'en-US': {
          // url is missing
          fileName: 'anAssets.png',
          contentType: 'image/png'
        }
      }
    }
  }
  expect(() => assertPayload({
    entries: [],
    locales: [],
    contentTypes: [],
    assets: [brokenAsset, brokenAsset]
  })
  ).toThrowErrorMatchingSnapshot()
})
