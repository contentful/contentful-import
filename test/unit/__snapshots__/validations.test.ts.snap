// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`payload validation fails if given invalid tag 1`] = `"tags:1 - Get further details in the error log file"`;

exports[`payload validation should fail when given an invalid data 1`] = `"assets:2 - Get further details in the error log file"`;
